src/Utilities.cpp

```diff
@@ -276,6 +277,74 @@ void lagrangeInterp(
   }
 }
 
+double brentRoot(
+  double lowerBound,
+  double upperBound,
+  double (*func)(double),
+  double epsilon) {
+    double counterPoint = lowerBound;
+    double currentPoint = upperBound;
+    double counterFunc = func(counterPoint);
+    double currentFunc = func(currentPoint);
+    if (counterFunc * currentFunc > 0.0) {
+      throw std::invalid_argument("Function values at the boundaries have the same sign.");
+    }
+    if (fabs(counterFunc) < fabs(currentFunc)) {
+      std::swap(counterPoint, currentPoint);
+      std::swap(counterFunc, currentFunc);
+    }
+
+    double previousPoint = counterPoint;
+    double previousFunc = counterFunc;
+    double evenOlderPoint = previousPoint;
+    double nextPoint;
+    double nextFunc;
+    int iteration = 0;
+    bool bisected = true;
+
+    do {
+      // Inverse quadratic interpolation
+      if (counterFunc != previousFunc && counterFunc != currentFunc) {
+        nextPoint = (counterPoint * currentFunc * previousFunc) / ((counterFunc - currentFunc) * (counterFunc - previousFunc));
+        nextPoint += (currentPoint * counterFunc * previousFunc) / ((currentFunc - counterFunc) * (currentFunc - previousFunc));
+        nextPoint += (previousPoint * currentFunc * counterFunc) / ((previousFunc - counterFunc) * (previousFunc - currentFunc));
+      }
+      // Secant method
+      else {
+        nextPoint = currentPoint - currentFunc * (currentPoint - counterPoint) / (currentFunc - counterFunc);
+      }
+
+      // Bisection method
+      if (((currentPoint - nextPoint) * (nextPoint - (3 * counterPoint + currentPoint) / 4) < 0) ||
```

Comment:
It would still be nice to have the comments you mentioned potentially adding in here about what the various if-checks in here are doing.

Reply:
But they're also really clearly enumerated in https://en.wikipedia.org/wiki/Brent%27s_method, so no big deal if you don't think it's worth adding.

Reply:
I think I'm just going to leave the comment out. If explanation if needed it's better to just point people to either a lecture or wikipedia.